examples/asr/asr_webapp/model_api.py

```diff
@@ -0,0 +1,131 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import contextlib
+import glob
+import os
+
+import torch
+
+import nemo.collections.asr as nemo_asr
+from nemo.utils import logging, model_utils
+
+# setup AMP (optional)
+if torch.cuda.is_available() and hasattr(torch.cuda, 'amp') and hasattr(torch.cuda.amp, 'autocast'):
+    logging.info("AMP enabled!\n")
+    autocast = torch.cuda.amp.autocast
+else:
+
+    @contextlib.contextmanager
+    def autocast():
+        yield
+
+
+MODEL_CACHE = {}
+
+# Special tags for fallbacks / user notifications
+TAG_ERROR_DURING_TRANSCRIPTION = "<ERROR_DURING_TRANSCRIPTION>"
+
+
+def get_model_names():
+    # Populate local copy of models
+    local_model_paths = glob.glob(os.path.join('models', "**", "*.nemo"), recursive=True)
+    local_model_names = list(sorted([os.path.basename(path) for path in local_model_paths]))
+
+    # Populate with pretrained nemo checkpoint list
+    nemo_model_names = set()
+    for model_info in nemo_asr.models.ASRModel.list_available_models():
+        for superclass in model_info.class_.mro():
+            if 'CTC' in superclass.__name__ or 'RNNT' in superclass.__name__:
+                nemo_model_names.add(model_info.pretrained_model_name)
+                break
+    nemo_model_names = list(sorted(nemo_model_names))
+    return nemo_model_names, local_model_names
+
+
+def initialize_model(model_name):
+    # load model
+    if model_name not in MODEL_CACHE:
+        if '.nemo' in model_name:
+            # use local model
+            model_name_no_ext = os.path.splitext(model_name)[0]
+            model_path = os.path.join('models', model_name_no_ext, model_name)
+
+            # Extract config
+            model_cfg = nemo_asr.models.ASRModel.restore_from(restore_path=model_path, return_config=True)
+            classpath = model_cfg.target  # original class path
+            imported_class = model_utils.import_class_by_path(classpath)  # type: ASRModel
+            logging.info(f"Restoring local model : {imported_class.__name__}")
+
+            # load model from checkpoint
+            model = imported_class.restore_from(restore_path=model_path, map_location='cpu')  # type: ASRModel
+
+        else:
+            # use pretrained model
+            model = nemo_asr.models.ASRModel.from_pretrained(model_name, map_location='cpu')
+
+        model.freeze()
+
+        # cache model
+        MODEL_CACHE[model_name] = model
+
+    model = MODEL_CACHE[model_name]
+    return model
+
+
+def transcribe_all(filepaths, model_name, use_gpu_if_available=True):
+    # instantiate model
+    if model_name in MODEL_CACHE:
+        model = MODEL_CACHE[model_name]
+    else:
+        model = initialize_model(model_name)
+
+    if torch.cuda.is_available() and use_gpu_if_available:
+        model = model.cuda()
+
+    # transcribe audio
+    logging.info("Begin transcribing audio...")
+    try:
+        with autocast():
```

Comment:
Does it always use AMP? Is it useful to make it an option or it does not matter?

Reply:
AMP is used only if torch amp is possible - it defaults to False for all hardware that doesnt support mixed precision. If it exists, it should be safe to use for inference - it preserves memory as well as improves speed of transcription.